benchmarkLive: a new `LiveTestWidgetsFlutterBindingFramePolicy` for benchmark on device

CareF opened the pull request:
## Description

Currently we have 5 `LiveTestWidgetsFlutterBindingFramePolicy`s, out of which, only `fullyLive` can or is intended to be used for running test on a device that respects the frame scheduling from engine. The problem is that it's still responding to `pump`s from both user calls of a test app or from some test gestures, which leads to unexpected frame requests (and therefore extra delay for awaiting the frame) besides those from engine and framework animation. This prevents a faithful simulation for performance test cases running on a device. 

For a test running without a device (i.e. `AutomatedTestWidgetsFlutterBinding`) `pump` is inevitable in coding a test case and also in implementing our API (e.g. in `flingFrom ` or `handlePointerEventRecord`) for driving a test app because before/after pump the location of a widget may change, which will change the hit test result and change the behavior to an input. However, for test running on a device (i.e. `LiveTestWidgetsFlutterBinding `) adding `pump`s will interfere with the engine frame scheduling, results in an unfaithful performance test result. 

For integrating test purposes we want to have a frame policy that's close to what a real app is doing: respect the frame request from the engine and from the framework, but ignore any extra frame request from the driving the test. 

Use demo is also the test case listed below, where we can see that **with the new frame policy flag, the delay for simulating gestures are significantly reduced**. 

<s>It's currently still labeled draft because the test case about frame count seems platform dependent and is flaky on cirrus</s>

## Related Issues

Resolves #60739

## Tests

I added the following tests:

- `dev/devicelab/bin/tasks/frame_policy_delay_test_android.dart` from `dev/integration_tests/simple_scroll/test/frame_policy.dart`. 

A typical result for the device lab test running locally on Pixel4 is: 
```json
{
  "success": true,
  "data": {
    "average_delay_fullyLive_millis": 9.096860655737704,
    "average_delay_benchmarkLive_millis": 1.986188524590164,
    "90th_percentile_delay_fullyLive_millis": 17.468,
    "90th_percentile_delay_benchmarkLive_millis": 4.105
  },
  "benchmarkScoreKeys": [
    "average_delay_fullyLive_millis",
    "average_delay_benchmarkLive_millis",
    "90th_percentile_delay_fullyLive_millis",
    "90th_percentile_delay_benchmarkLive_millis"
  ]
}
```
on Moto G4 is:
```json
{
  "success": true,
  "data": {
    "average_delay_fullyLive_millis": 10.99943442622951,
    "average_delay_benchmarkLive_millis": 4.632606557377049,
    "90th_percentile_delay_fullyLive_millis": 22.621,
    "90th_percentile_delay_benchmarkLive_millis": 8.177
  },
  "benchmarkScoreKeys": [
    "average_delay_fullyLive_millis",
    "average_delay_benchmarkLive_millis",
    "90th_percentile_delay_fullyLive_millis",
    "90th_percentile_delay_benchmarkLive_millis"
  ]
}
```

This `simple_scroll` may also be used for other scrolling tests. I avoid use `complex_layout` because that one is more noisy due to a complex layout. 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

CareF committed:
add benchmarkLive flag and tests

CareF commented:
The test passes on my local machine, but I'm guessing the CI system when running `flutter test` with `LiveTestWidgetsFlutterBinding` it has a different refresh rate.

dnfield commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -1264,6 +1264,12 @@ enum LiveTestWidgetsFlutterBindingFramePolicy {
+  /// Ignore any request from pump.
+  ///
+  /// This is used for running the test on a device, where scheduling of new
+  /// frames respect what the engine and the device needed.
```
It would help to explain how this differs from benchmark and fullyLive, and when I'd want to use this instead of one of those.

CareF replied:
`benchmark` is actually very different: it's ignoring all animations. But yeah I should update the docs, and restructure the two parts from https://api.flutter.dev/flutter/flutter_test/LiveTestWidgetsFlutterBindingFramePolicy-class.html and
https://api.flutter.dev/flutter/flutter_test/LiveTestWidgetsFlutterBinding/framePolicy.html . But before that I still would like to hear if the general idea is reasonable.

dnfield commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -1489,6 +1496,10 @@ class LiveTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
     assert(_pendingFrame == null);
+    if (framePolicy == LiveTestWidgetsFlutterBindingFramePolicy.benchmarkLive) {
+      // Ignore all pumps and just wait.
+      return Future<void>.delayed(duration?? Duration.zero);
```
nit: space between `duration ??`.

Why are we awaiting a delay of the duration before stting a timer for the duration again below? Won't that mean we get double the duration expected?

CareF replied:
It's an early return. When the policy is `benchmarkLive`, the below part is not executed.

dnfield replied:
Ahh sorry, I missed the scope of the return here.

liyuqian commented on packages/flutter_test/test/live_binding_test.dart:
```diff
@@ -16,4 +18,93 @@ void main() {
+    ];
+    final List<Duration> delays = await tester.handlePointerEventRecord(records);
+
+    // This will typically be 113 +- 1 for fullyLive flag.
```
What's the number here for other flags, on host unit tests and on device?

CareF replied:
All other flags don't response to frame requests from the engine, including the benchmark, and they basically don't track animation unless manually pump-ed. So these numbers are not relevant for other flags. 
As for the number on host and on device, that's the issue why I'm getting flaky result. The number here comes from my local test on my macbook, and it seems on Cirrus they get different number, varying from 60+- ~90+-. Is there better place to test these number?

dnfield commented on packages/flutter_test/test/live_binding_test.dart:
```diff
@@ -16,4 +18,93 @@ void main() {
     await gesture.addPointer();
   });
+
+  binding.framePolicy = LiveTestWidgetsFlutterBindingFramePolicy.fullyLive;
```
This won't work. Tests might be executed in random order, and the code "on top of them" isn't guaranteed to run inside the test itself.

You either have to set the frame policy in the test and have that just work, or create separate test files for different policies.

You _might_ get this working if you use `group`s.

CareF commented:
After discussion with @liyuqian we feel it's probably better to move the test case to device lab and measure the delayed time reported by `handlePointerEventRecord` as a metric (average, 90th percentile) and use this as if it's a performance test. And the frame count is not very different anyway for `fullyLive` and `benchmarkLive` anyway and we sill ignore it.

@dnfield What do you think?

dnfield commented:
Sgtm

CareF commented:
A typical result for the device lab test running locally is: 
```json
{
  "success": true,
  "data": {
    "average_delay_fullyLive_millis": 11.110319672131148,
    "average_delay_benchmarkLive_millis": 3.6040245901639345,
    "90th_percentile_delay_fullyLive_millis": 19.567,
    "90th_percentile_delay_benchmarkLive_millis": 9.01
  },
  "benchmarkScoreKeys": [
    "average_delay_fullyLive_millis",
    "average_delay_benchmarkLive_millis",
    "90th_percentile_delay_fullyLive_millis",
    "90th_percentile_delay_benchmarkLive_millis"
  ]
}
```

CareF commented:
There're already too many things. I'll open another PR to deal with the docs after this lands for docs. The major purpose would be to combine the two separate docs explaining the flag, but it will also be used to clarify the usage.

liyuqian commented:
I can no longer see the "Draft" status. Does "It's currently still labeled draft because the test case about frame count seems platform dependent and is flaky on cirrus" in the PR description still aplply?

liyuqian commented on dev/integration_tests/simple_scroll/test/frame_policy.dart:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
To reduce the number of new files introduced, can we put this under `macrobenchmarks/test`, and create a new page in `macrobenchmarks` for the purpose of `simple_scroll`?

CareF replied:
I don't feel comfortable doing that because it's technically not a performance benchmark but more like a functioning test. But if you insist I'll make the change.

liyuqian replied:
The reporting of metrics like `average_delay_fullyLive_millis` and `average_delay_benchmarkLive_millis` makes me feel that this is more or less a performance test. It's testing the performance of the input events simulation which justifies `benchmarkLive` (i.e., everything would still run if we don't have `benchmarkLive`, but the simulation performance would be bad and inaccurate). Such performance metrics also prevent future accidents that suddenly make our simulation much less accurate.

CareF replied:
OK. Then I'll wait this one until that renaming PR lands to avoid conflicts.

CareF replied:
Done.

liyuqian commented on dev/devicelab/lib/tasks/perf_tests.dart:
```diff
@@ -250,6 +250,50 @@ TaskFunction createsMultiWidgetConstructPerfTest() {
+      ) as Map<String, dynamic>;
+      final Map<String, dynamic> fullLiveData = data['fullyLive'] as Map<String, dynamic>;
+      final Map<String, dynamic> benchmarkLiveData = data['benchmarkLive'] as Map<String, dynamic>;
+      final Map<String, dynamic> dataFormated = <String, dynamic>{
```
It would be nice to put your local results in the PR description to set the expectation of those two different flags.

CareF replied:
Done.

CareF commented:
@liyuqian It shouldn't be considered draft now. That's out-of-date because I removed the unit test and moved it to a device lab test. 
Although to enable merge I'll need to wait for package updates.

CareF committed:
update pubspec

CareF commented on dev/benchmarks/macrobenchmarks/test/frame_policy.dart:
```diff
@@ -0,0 +1,105 @@
+  testWidgets(
+    'Frame Counter and Input Delay for benchmarkLive',
+    (WidgetTester tester) async {
+      await tester.pumpWidget(MaterialApp(home: Scaffold(body: SimpleScroll())));
```
By doing so I'm avoiding extra operation to select the page. It might be a good idea to do the same for other pages with E2E test version, or to put this together with the util code in #61509

CareF commented on dev/benchmarks/macrobenchmarks/test_driver/frame_policy_test.dart:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
This file is modified from https://github.com/flutter/plugins/blob/master/packages/e2e/lib/e2e_driver.dart

CareF committed:
update docs

liyuqian commented on dev/benchmarks/macrobenchmarks/test/frame_policy.dart:
```diff
@@ -0,0 +1,105 @@
+void _reportResult(String eventName, int frameCount, List<Duration> delays) {
+  assert(delays.length > 1);
+  final List<int> delayedInMicro = delays.map<int>(
+    (Duration delay) => -delay.inMicroseconds,
```
This looks strange. Does `handlePointerEventRecord` mostly return all negative delays? Shall we use `abs` just in case of positive delays? We might need to add some documentations to `handlePointerEventRecord` to describe the meaning of positive and negative delays, and set the expectation if most of them are negative or positive.

CareF replied:
Yes mostly negative. It's positive only when `Future.delayed` complete earlier than the set time. My local test shows it's 0~3 out of 200+ events with tiny positive values < 100us. I don't think I should use `abs` because this list is only shown in debug mode and `abs` makes less use of the info. But it does make sense to change the behavior of `handlePointerEventRecord` for the major result being positive since we are calling it delay, and add docs to that.

CareF replied:
Update about the number. When putting this on a more complex environment, the fluctuation becomes larger, and positives can go as much as ~10% and up to 500us. Still tiny compare to frame interval though.

liyuqian commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -1264,6 +1264,20 @@ enum LiveTestWidgetsFlutterBindingFramePolicy {
   /// application has "settled".
   benchmark,
+
+  /// Ignore any request from pump.
```
`Ignore any request from pump but respect other requests to schedule a frame.`

CareF committed:
modify according to liyuqian@

CareF committed:
update handlePointerEventRecord doc

CareF committed:
using e2e 0.6.1

CareF commented:
And I also just realized I should use flutter/plugins#2873 this new feature of reporting result in e2e to implement this rather than using timeline. Added a new commit. @liyuqian

CareF commented:
I may also need to update the macrobenchmark/readme.md, which is also the case for #61388. That will be done on whichever of the two PR who lands later.

liyuqian reviewed (approved):
LGTM
